pkg/storage/cacher.go

```diff
@@ -768,6 +770,10 @@ func (c *cacheWatcher) add(event *watchCacheEvent) {
 		c.forget(false)
 		c.stop()
 	}
+
+	if *timeout = *timeout - time.Since(startTime); *timeout < 0 {
```

Comment:
I would probably extract this and the timerPool bit to a separate function / object, just for clarity & to make testing easier. Up to you.